omod/src/main/java/org/openmrs/module/pihcore/fragment/controller/dashboardwidgets/RiskFactorsFragmentController.java

```diff
@@ -0,0 +1,136 @@
+package org.openmrs.module.pihcore.fragment.controller.dashboardwidgets;
+
+import org.apache.commons.lang.StringUtils;
+import org.codehaus.jackson.JsonNode;
+import org.openmrs.Concept;
+import org.openmrs.Encounter;
+import org.openmrs.Obs;
+import org.openmrs.Patient;
+import org.openmrs.PatientProgram;
+import org.openmrs.Program;
+import org.openmrs.api.ConceptService;
+import org.openmrs.api.ObsService;
+import org.openmrs.api.PatientService;
+import org.openmrs.api.ProgramWorkflowService;
+import org.openmrs.module.appframework.domain.AppDescriptor;
+import org.openmrs.module.emrapi.patient.PatientDomainWrapper;
+import org.openmrs.ui.framework.annotation.FragmentParam;
+import org.openmrs.ui.framework.annotation.SpringBean;
+import org.openmrs.ui.framework.fragment.FragmentConfiguration;
+import org.openmrs.ui.framework.fragment.FragmentModel;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+public class RiskFactorsFragmentController {
+    private final Logger log = LoggerFactory.getLogger(getClass());
+
+    public void controller(@SpringBean("patientService") PatientService patientService,
+                           @SpringBean("conceptService") ConceptService conceptService,
+                           @SpringBean("obsService") ObsService obsService,
+                           @SpringBean("programWorkflowService") ProgramWorkflowService programWorkflowService,
+                           @FragmentParam("app") AppDescriptor app,
+                           FragmentConfiguration config,
+                           FragmentModel model) {
+
+        Object patientConfig = config.get("patient");
+        if (patientConfig == null) {
+            patientConfig = config.get("patientId");
+        }
+        Patient patient = null;
+        if (patientConfig != null) {
+            if (patientConfig instanceof Patient) {
+                patient = (Patient) patientConfig;
+            } else if (patientConfig instanceof PatientDomainWrapper) {
+                patient = ((PatientDomainWrapper) patientConfig).getPatient();
+            } else if (patientConfig instanceof Integer) {
+                patient = patientService.getPatient((Integer) patientConfig);
+            } else if (patientConfig instanceof String) {
+                patient = patientService.getPatientByUuid((String) patientConfig);
+            }
+        }
+        if (patient == null) {
+            throw new IllegalArgumentException("No patient found. Please pass patient into the configuration");
+        }
+
+        model.put("app", app);
+        model.put("patient", patient);
+
+        JsonNode node = app.getConfig().get("concepts");
```

Comment:
I'm a little confused about this configuration.  I would expect to see this kind of configuration if we named this fragment something more generic (eg. ObsByEncounterFragmentController), which we could use for various use cases, one of which is to display Risk Factors.  But I don't really understand the configuration if we are naming this controller so specifically.  Can we rename these to something more generic please?  Or if these are hard-coded in a way that is only appropriate for risk factors, can we get rid of the configuration?

Reply:
Yeah, this looks like just a basic Obs-In-Program widget, and we could just call it something like that, but @cioan let me know if we are missing something.